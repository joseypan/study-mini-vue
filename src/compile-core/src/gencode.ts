export const generate = (ast) => {
  const context = createGenerateContext();
  const functionName = "render";
  const argumentsList = ["_ctx", "_cache"];
  const argument = argumentsList.join(", ");
  context.push(`return function ${functionName}(${argument}){`);
  genNode(ast, context);
  context.push(`}`);
  return {
    code: context.code,
  };
};
/**
 * 描述：获取节点的渲染内容
 * @param { any } ast
 * @param { {code:string,push:(source:string)=>void} } context
 * @return
 */
const genNode = (ast, context) => {
  const { push } = context;
  const content = ast.gencodeNode.content;
  push(`return ${content}`);
};

/**
 * 描述：创建一个generate上下文环境
 * @param {  }
 * @return
 */
const createGenerateContext = () => {
  const context = {
    code: "",
    push: (source) => {
      context.code += source;
    },
  };
  return context;
};
