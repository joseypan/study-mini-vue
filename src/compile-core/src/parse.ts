import { NodeTypes } from "./ast";
enum TagType {
  /*
   * 描述：tag开始标签
   */
  TAGSTART = "TAGSTART",
  /*
   * 描述：tag结束标签
   */
  TAGEND = "TAGEND",
}
/**
 * 描述：对模板字符串进行解析
 * @param { string } content 传入的内容
 * @return
 */
export const baseParese = (content: string) => {
  const context = createParseContext(content); //创建一个上下文环境
  return createRoot(parseChildren(context, []));
};
/**
 * 描述：创建根节点
 * @param { array } children 根内容的子节点
 * @return {children:array} 子节点
 */
const createRoot = (children) => {
  return {
    children,
  };
};
/**
 * 描述：判断children是否都被处理完成
 * @param {  }
 * @return
 */
const isEnd = (context: any, ancestors: { tag: string }[]) => {
  // 终止的情况：1、context.source已经不存在了 2、已经到了结束的标签
  const s = context.source;
  if (s.startsWith("</")) {
    //把当前项和ancestors的最后一项对比，看是否是一致的，就可以看是否匹配
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (sourceStartsWithTag(s, tag)) {
        return true;
      }
    }
  }
  return !s;
};
/**
 * 描述：解析子节点
 * @param {  }
 * @return
 */
const parseChildren = (context, ancestors) => {
  const nodeList: any = [];
  let source = context.source;
  let node;
  // 这里需要将所有的children都进行解析，所以得用一个while循环来解决
  while (!isEnd(context, ancestors)) {
    if (source.startsWith("{{")) {
      node = parseInterpolation(context);
    } else if (source.startsWith("<")) {
      //这里主要是去除结束标签的问题
      if (/[a-z]/i.test(source[1])) {
        node = parseElement(context, ancestors);
      }
    }
    if (!node) {
      node = parseText(context);
    }
    nodeList.push(node);
    source = context.source;
  }
  return nodeList;
};

/**
 * 描述：解析表达式
 * @param {  }
 * @return
 */
const parseInterpolation = (context) => {
  //目前要对{{message}} 进行解析,需要将message提取出来,思路是先匹配}找到其索引作为endIndex，然后从index+2到endIndex进行截取即可
  const openDelimitre = "{{";
  const closeDelimitre = "}}";
  const endIndex = context.source.indexOf(closeDelimitre);
  context.source = advanceBy(context, openDelimitre.length);
  const contentRaw = parseTextData(context, endIndex - closeDelimitre.length);
  const content = contentRaw.trim();
  context.source = advanceBy(context, closeDelimitre.length);
  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  };
};
const advanceBy = (context: any, length: number) => {
  return context.source.slice(length);
};
const createParseContext = (content: string) => {
  return {
    source: content,
  };
};

/**
 * 描述：解析element的逻辑
 * @param { any } context 上下文环境
 * @return
 */
const parseElement = (context: any, ancestors) => {
  const node = parseTag(context, TagType.TAGSTART);
  if (!node) return;
  ancestors.push(node);
  node.children = parseChildren(context, ancestors);
  ancestors.pop();
  if (sourceStartsWithTag(context.source, node.tag)) {
    parseTag(context, TagType.TAGEND);
  } else {
    throw new Error(`lack end tag:${node.tag}`);
  }
  return node;
};
const sourceStartsWithTag = (source, tag) => {
  return source.startsWith("</") && source.slice(2, 2 + tag.length) === tag;
};
const parseTag = (context: any, type: TagType) => {
  // 解析tag
  // 删除解析出来的内容
  const tagReg = /^<\/?([a-z]*)/i;
  const execList: any = tagReg.exec(context.source);
  const tag = execList[1];
  context.source = advanceBy(context, execList[0].length);
  context.source = advanceBy(context, 1);
  if (type === TagType.TAGEND) return;
  return {
    type: NodeTypes.ELEMENT,
    tag: tag,
    children: [],
  };
};
/**
 * 描述：解析text
 * @param { any } context 上下文
 * @return
 */
const parseText = (context) => {
  //这里需要区分，如果是{{开头或者是<开头，则不应该到字符串结尾处
  let endIndex = context.source.length;
  const endCharater = ["{{", "<"];
  for (let i = 0; i < endCharater.length; i++) {
    const index = context.source.indexOf(endCharater[i]);
    if (index !== -1 && index < endIndex) {
      //说明需要重新定位截取位置
      endIndex = index;
    }
  }
  // 解析
  const content = parseTextData(context, endIndex);
  // 推进
  return {
    type: NodeTypes.TEXT,
    content,
  };
};

/**
 * 描述：处理text内容
 * @param {  }
 * @return
 */
const parseTextData = (context, length) => {
  const content = context.source.slice(0, length);
  context.source = advanceBy(context, content.length);
  return content;
};
