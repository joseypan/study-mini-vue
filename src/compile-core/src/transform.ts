import { NodeTypes } from "./ast";
/**
 * @FileDescription:处理编译阶段的transform过程
 * @Author: 潘旭敏
 * @Date: 2022-11-10
 * @LastEditors: 潘旭敏
 * @LastEditTime:2022-11-10 00:20
 */
export const transform = (root, options) => {
  const context = createTransformContext(root, options);
  //遍历整个ast树，然后找到对应的节点进行处理
  tranversTree(root, context);
};
/**
 * 描述：创建transform的上下文环境
 * @param {  }
 * @return
 */
const createTransformContext = (root, options) => {
  return {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };
};

/**
 * 描述：遍历树结构
 * @param { any } root 根节点
 * @param { {root:any,nodeTransforms:any[]} } context 上下文环境
 * @return
 */
const tranversTree = (root, context) => {
  if (!root) return null;
  const stack = [root];
  while (stack.length) {
    const curNode = stack.pop();
    if (curNode === null) {
      const node = stack.pop();
      //这里需要对节点类型进行判断，可能要对不同的节点类型进行不同的操作，但是如果把所有的判断都写在这，就把程序写死了，而且很臃肿
      for (let i = 0; i < context.nodeTransforms.length; i++) {
        context.nodeTransforms[i](node);
      }
    } else {
      //这里有个坑，curNode不一定有children属性
      const children = curNode.children || [];
      for (let i = children.length - 1; i >= 0; i--) {
        stack.push(children[i]);
      }
      stack.push(curNode);
      stack.push(null);
    }
  }
};
