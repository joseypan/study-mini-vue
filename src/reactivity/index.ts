export function add(a:any,b:any){
    return a+b;
}