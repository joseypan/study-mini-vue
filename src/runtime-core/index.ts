// export
export { h } from "./h";

export { renderSlots } from "./helpers/renderSlots";

export { createTextVNode } from "./vnode";

export { getCurrentInstance } from "./component";

export { provider, inject } from "./apiInject";

export { nextTick } from "./schedule";
