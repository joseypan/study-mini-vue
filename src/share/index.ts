/**
 * @FileDescription:用于公共处理的方法
 * @Author: 潘旭敏
 * @Date: 2022-08-09
 * @LastEditors: 潘旭敏
 * @LastEditTime:2022-08-09 00:02
 */
export const isObject = (data: unknown) => {
  return data !== null && typeof data === "object";
};
/**
 * 描述：判断值是否发生改动
 * @param { any } rawValue newValue
 * @return boolean
 */
export function hasChanged(rawValue, newValue) {
  return Object.is(rawValue, newValue);
}
