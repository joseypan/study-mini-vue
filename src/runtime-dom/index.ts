import { createRenderer } from "../runtime-core/renderer";

// 这里用来实现createElement patchProps insert方法的具体实现
function createElement(type) {
  return document.createElement(type);
}

function patchProps(el, key, val) {
  // 事件的格式是on+大写字母，当属性是以此开头的时候，默认是一个事件
  const isEvent = (key) => /^on[A-Z]/.test(key);
  if (isEvent(key)) {
    // 需要绑定在当前元素上，也就是vnode的el上
    const event = key.slice(2).toLocaleLowerCase();
    el.addEventListener(event, val);
  } else {
    el.setAttribute(key, val);
  }
}

function insert(el, container) {
  // 挂载
  container.append(el);
}

const renderer: any = createRenderer({
  createElement,
  patchProps,
  insert,
});

export function createApp(...arg) {
  return renderer.createApp(...arg);
}

export * from "../runtime-core";
