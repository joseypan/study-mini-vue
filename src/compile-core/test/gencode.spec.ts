import { transform } from "./../src/transform";
import { generate } from "../src/gencode";
import { baseParese } from "./../src/parse";
describe("gencode", () => {
  it.only("string", () => {
    const ast = baseParese("hi");
    transform(ast);
    const { code } = generate(ast);
    const expectResult = `
    return function render(_ctx, _cache, $props, $setup, $data, $options) {
      return "hi"
    }
    `;
    expect(code).toMatchSnapshot(expectResult);
  });
});
