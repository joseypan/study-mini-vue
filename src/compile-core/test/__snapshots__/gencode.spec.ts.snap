// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gencode string: 
    return function render(_ctx, _cache, $props, $setup, $data, $options) {
      return "hi"
    }
     1`] = `"return function render(_ctx, _cache){return hi}"`;
